### ISSUE TITLE
ENH: need matlab compatible min and max for complex numbers

### ISSUE THREAD

**Issue by jakirkham** [numpy/numpy#8151]
If I take the minimum or maximum of complex numbers, I would expect that the magnitude is used. However, I find the real component is used instead. Here is a simple example. My expectation would be max and min are reversed in this case.

```python
In [1]: import numpy

In [2]: a = numpy.array([1+1j, 1.2])

In [3]: numpy.abs(a)
Out[3]: array([ 1.41421356,  1.2       ])

In [4]: numpy.real(a)
Out[4]: array([ 1. ,  1.2])

In [5]: numpy.min(a)
Out[5]: (1+1j)

In [6]: numpy.max(a)
Out[6]: (1.2+0j)
```

**Comment by pv** [comment]
It's lexicographic order. I'm not sure I agree absolute value would be less surprising.

**Comment by mhvk** [comment]
From http://math.stackexchange.com/questions/310931/comparing-complex-numbers#310941, I think this boils down to asking whether complex numbers can be ordered at all, with the answer being that, no, generally they cannot. So, the logical consequence would be not to allow finding a minimum at all. Indeed, this is the case in python proper:
```
min(1+1j, 1+2j)
TypeError: unorderable types: complex() < complex()
```

Indeed, comparison itself is not defined in python, but defined in numpy:
```
1+1j < 1+2j
# TypeError: unorderable types: complex() < complex()
np.array(1+1j) < np.array(1+2j)
# True
```

Anyway, not something one can change lightly.

**Comment by ahaldane** [comment]
For what it's worth, Matlab uses the magnitude here. Pure-python and julia raise errors.

**Comment by pv** [comment]
Yes, complex number comparison is ill defined. Lexical order is 
compatible with addition, but not multiplication. Magnitude is not total 
ordering.
.
But I would leave it as it is, and leave it to the user to know what 
they're doing when comparing complex numbers.

**Comment by pv** [comment]
Suggest closing (behavior appears intended, changing probably disruptive).

**Comment by charris** [comment]
Yep, the sort order is lexicographic.

**Comment by jakirkham** [comment]
So my follow-up question was could we have `min` and `max` functions (or arguments to such functions that match the MATLAB behavior. As someone who works with scientist that are use to MATLAB, it is hard to sell them on Python when they find expectations like these are not met. Carrying around utility function like this on my end becomes silly and often redundant. Would it be possible to have some like this in `numpy` proper? I'm happy to ask this in a new issue if you would rather.

**Comment by charris** [comment]
I don't think it would be too difficult to make such functions, but with different names than max/min. A PR would be welcome.

**Comment by jakirkham** [comment]
Ok, could I ask that we reopen this issue so that we can keep track of it? I don't have time for such a PR now, but would be willing to do it in the near future.

**Comment by charris** [comment]
OK, I reopened with a `Wish List` label.

**Comment by hmaarrfk** [comment]
I don't know if there is still interest here. But I would be interested in such a function too.

Would it be called: `np.max_abs`?

Application note: In optics, it is sometimes useful to find the point of maximum intensity, but to propagate your field as a complex number.

**Comment by charris** [comment]
Signed ints are a bit of a problem for abs, as the most negative values are invariant. If we stick to inexact types, no problem. Maybe always return an unsigned type, but then there are unexpected promotions. Maybe just warn when problem encountered?

**Comment by hmaarrfk** [comment]
For integers, maybe it should be implemented as minmax? The the check for the most negative value becomes order 1

**Comment by hmaarrfk** [comment]
I get it. but max_abs could do something like
pseudocode

```
if dtype == signed int
    min, max = np.minmax(a)
    if min == -128
         warn
    else
         return maximum(abs(min), abs(max))
```

**Comment by leofang** [comment]
related: #2004
